src/lem/circuit.rs

```diff
@@ -0,0 +1,899 @@
+                            }
+                            LEMOP::Hash3(img, tag, preimg) => {
+                                hash_helper!(img, tag, preimg, SlotType::Hash3);
+                            }
+                            LEMOP::Hash4(img, tag, preimg) => {
+                                hash_helper!(img, tag, preimg, SlotType::Hash4);
+                            }
+                            LEMOP::Unhash2(preimg, img) => {
+                                unhash_helper!(preimg, img, SlotType::Hash2);
+                            }
+                            LEMOP::Unhash3(preimg, img) => {
+                                unhash_helper!(preimg, img, SlotType::Hash3);
+                            }
+                            LEMOP::Unhash4(preimg, img) => {
+                                unhash_helper!(preimg, img, SlotType::Hash4);
+                            }
+                            LEMOP::Null(tgt, tag) => {
+                                let allocated_tgt = LEM::allocate_ptr(
+                                    g.cs,
+                                    &LEM::zptr_from_mptr(tgt, g.frame, g.store)?,
+                                    tgt.name(),
+                                    &mut g.allocated_ptrs,
+                                )?;
+                                let allocated_tag =
+                                    g.global_allocator.get_or_alloc_num(g.cs, tag.to_field())?;
+
+                                // Constrain tag
+                                implies_equal(
+                                    &mut g
+                                        .cs
+                                        .namespace(|| format!("implies equal for {tgt}'s tag")),
+                                    &concrete_path,
+                                    allocated_tgt.tag(),
+                                    &allocated_tag,
+                                )
+                                .with_context(|| {
+                                    format!("couldn't enforce implies equal for {tgt}'s tag")
+                                })?;
+
+                                // Constrain hash
+                                implies_equal_zero(
+                                    &mut g.cs.namespace(|| {
+                                        format!("implies equal zero for {tgt}'s hash")
+                                    }),
+                                    &concrete_path,
+                                    allocated_tgt.hash(),
+                                )
+                                .with_context(|| {
+                                    format!("couldn't enforce implies equal zero for {tgt}'s hash")
+                                })?;
+                            }
+                            _ => todo!(),
+                        }
+                    }
+                    recurse(rest, concrete_path, path, slots_count, g)
```

Comment:
Shouldn't this have constructed and passed a `new_path`? I would expect this to lead to constraint-system path/namespace conflicts under some (probably not currently tested) conditions.

Reply:
This is for the `Seq` case, which continues on the same path